# HG changeset patch
# User Betterbird <betterbird@betterbird.eu>
# Date 1788286832 -7200
# Parent  538c141e2260edcd4aa8d741fba5dcf769958c97
Feature: Follow-up to bug 2000453: Add combination of MRU and alphabetic for recent folders.

Based on the original idea in https://phabricator.services.mozilla.com/D277216 which wasn't accepted by the TB folks as they preferred a dumber solution.
Betterbird already shipped this solution in version 140.

diff --git a/mail/components/preferences/general.inc.xhtml b/mail/components/preferences/general.inc.xhtml
--- a/mail/components/preferences/general.inc.xhtml
+++ b/mail/components/preferences/general.inc.xhtml
@@ -798,10 +798,11 @@
       <hbox align="center" pack="start">
         <label control="recentSortOrder" data-l10n-id="recent-sort-order-label"/>
         <menulist id="recentSortOrder" preference="mail.folder_widget.recent_sort_order">
           <menupopup>
+            <menuitem data-l10n-id="recent-sort-order-combined" value="2"/>
+            <menuitem data-l10n-id="recent-sort-order-alphabetic" value="1"/>
             <menuitem data-l10n-id="recent-sort-order-mru" value="0"/>
-            <menuitem data-l10n-id="recent-sort-order-alphabetic" value="1"/>
           </menupopup>
         </menulist>
       </hbox>
 
diff --git a/mail/locales/en-US/messenger/preferences/preferences.ftl b/mail/locales/en-US/messenger/preferences/preferences.ftl
--- a/mail/locales/en-US/messenger/preferences/preferences.ftl
+++ b/mail/locales/en-US/messenger/preferences/preferences.ftl
@@ -590,8 +590,11 @@ recent-sort-order-mru =
 
 recent-sort-order-alphabetic =
     .label = Alphabetical
 
+recent-sort-order-combined =
+    .label = { recent-sort-order-mru.label } + { recent-sort-order-alphabetic.label }
+
 max-recent-label = Maximum number of recent folders:
     .accesskey = M
 
 account-hub-legend = Account hub
diff --git a/mailnews/base/content/folder-menupopup.js b/mailnews/base/content/folder-menupopup.js
--- a/mailnews/base/content/folder-menupopup.js
+++ b/mailnews/base/content/folder-menupopup.js
@@ -625,9 +625,40 @@
             : folderItem.folderName;
         }
 
         // If appropriate, sort the entries alphabetically.
+        let mruCount = 0;
         if (
+          Services.prefs.getIntPref(
+            "mail.folder_widget.recent_sort_order",
+            0
+          ) == 2
+        ) {
+          // Sort the special folders alphabetically, except for "Recent
+          // destinations" where a short list of the most recently used folders
+          // is placed first.
+          const totalCount = specialFoldersMap.length;
+          if (specialType == "recent") {
+            // Show about 20% of the total items, but at most 5, in the order of
+            // most recent use. Also show small lists consisting of up to five
+            // items in MRU order.
+            mruCount =
+              totalCount < 6
+                ? totalCount
+                : Math.min(Math.ceil(totalCount / 5), 5);
+          }
+          let mruFoldersMap = [];
+          if (mruCount) {
+            // Remove the MRU items from the main list.
+            mruFoldersMap = specialFoldersMap.splice(0, mruCount);
+          }
+          // Sort the main list alphabetically.
+          specialFoldersMap.sort((a, b) => a.label.localeCompare(b.label));
+          if (mruCount) {
+            // Re-insert MRU items at the front of the list.
+            specialFoldersMap.splice(0, 0, ...mruFoldersMap);
+          }
+        } else if (
           specialType != "recent" ||
           Services.prefs.getIntPref(
             "mail.folder_widget.recent_sort_order",
             0
@@ -636,9 +667,14 @@
           specialFoldersMap.sort((a, b) => a.label.localeCompare(b.label));
         }
 
         // Create entries for each of the special folders.
-        for (const folderItem of specialFoldersMap) {
+        specialFoldersMap.forEach((folderItem, index) => {
+          if (mruCount && index == mruCount) {
+            // Add a separator between the most recently used entries
+            // and the alphabetically sorted entries.
+            submenu.childWrapper.appendChild(this._buildSeparator());
+          }
           const attributes = {
             label: folderItem.label,
             tooltiptext: `${folderItem.folderPath} – ${folderItem.serverName}`,
             crop: "start",
@@ -646,9 +682,9 @@
           };
           submenu.childWrapper.appendChild(
             this._buildMenuItem(attributes, folderItem.folder)
           );
-        }
+        });
 
         if (specialFoldersMap.length === 0) {
           const stubItem = this._buildMenuItem(
             {
diff --git a/mailnews/mailnews.js b/mailnews/mailnews.js
--- a/mailnews/mailnews.js
+++ b/mailnews/mailnews.js
@@ -191,13 +191,13 @@ pref("mail.citation_color", "#007cff");
 #endif
 // Use CTE quoted-printable for mail bodies.
 pref("mail.strictly_mime", false);
 // The sort order for the entries in the 'Recent Destination' sub menu:
-// 0 by last used, 1 alphabetically.
-pref("mail.folder_widget.recent_sort_order", 0);
+// 0 by last used, 1 alphabetically, 2 combined.
+pref("mail.folder_widget.recent_sort_order", 2);
 // The maximum number of entries in the "Recent Destinations" menu of the
 // folder picker.
-pref("mail.folder_widget.max_recent", 10);
+pref("mail.folder_widget.max_recent", 20);
 // 0/1 (name param is encoded in a legacy way), 2(RFC 2231 only)
 // 0 the name param is never separated to multiple lines.
 pref("mail.strictly_mime.parm_folding", 1);
 pref("mail.label_ascii_only_mail_as_us_ascii", false);
